lib/AST/ModuleNameLookup.cpp

```diff
@@ -90,20 +95,111 @@ static bool updateOverloadSet(NamedCanTypeSet &overloads,
+    return true;
+  }
+}
+
+/// Returns whether it's okay to stop recursively searching imports, given that
+/// we found something, are using a NamedCanTypeSet to resolve overloads, and
+/// have the given \p resolutionKind.
+static bool canReturnEarly(const NamedCanTypeSet &overloads,
+                           ResolutionKind resolutionKind) {
+  return false;
+}
+
+namespace {
+using LookupCallbackTy = llvm::function_ref<
+    void(ModuleDecl *, ModuleDecl::AccessPathTy,
+         SmallVectorImpl<ValueDecl *> &)>;
+
+/// Encapsulates the work done for a recursive qualified lookup into a module.
+///
+/// The \c OverloadSetTy chosen determines how overloads are resolved.
 template <typename OverloadSetTy>
-static ResolutionKind recordImportDecls(LazyResolver *typeResolver,
-                                        SmallVectorImpl<ValueDecl *> &results,
-                                        ArrayRef<ValueDecl *> newDecls,
-                                        OverloadSetTy &overloads,
-                                        ResolutionKind resolutionKind) {
+class ModuleNameLookup {
+  ModuleLookupCache cache;
+  LookupCallbackTy callback;
+  LazyResolver *typeResolver;
+  bool respectAccessControl;
+
+  ModuleNameLookup() = default;
+
+  /// After finding decls by name lookup, filter based on the given
+  /// resolution kind and existing overload set and add them to \p results.
+  ///
+  /// Both \p overloads and \p resolutionKind are updated based on the new
+  /// declarations.
+  ///
+  /// \returns true if any decls were added to \p results.
+  bool recordImportDecls(SmallVectorImpl<ValueDecl *> &results,
+                         ArrayRef<ValueDecl *> newDecls,
+                         OverloadSetTy &overloads,
+                         ResolutionKind &resolutionKind);
+
+  /// Given a list of imports and an access path to limit by, perform a lookup
+  /// into each of them and record the results in \p decls, filtering based on
+  /// the given resolution kind and existing overload set.
+  void collectLookupResultsFromImports(
+      ArrayRef<ModuleDecl::ImportedModule> imports,
+      ModuleDecl::AccessPathTy accessPath,
+      SmallVectorImpl<ValueDecl *> &decls,
+      const DeclContext *moduleScopeContext,
+      OverloadSetTy &overloads, ResolutionKind resolutionKind);
+
+  /// Performs a qualified lookup into the given module and, if necessary, its
+  /// reexports, observing proper shadowing rules.
+  void lookupInModule(ModuleDecl *module,  ModuleDecl::AccessPathTy accessPath,
+                      SmallVectorImpl<ValueDecl *> &decls,
+                      ResolutionKind resolutionKind,
+                      const DeclContext *moduleScopeContext,
+                      ArrayRef<ModuleDecl::ImportedModule> extraImports = {});
+public:
+  /// Performs a qualified lookup into the given module and, if necessary, its
+  /// reexports, observing proper shadowing rules.
+  static void lookupInModule(ModuleDecl *startModule,
```

Comment:
At this point, seeing approx the same bunch o' parameters a few times, I'm starting to think about a missing concept--just a thought.